### ISSUE TITLE
Swift 3 API Design

### ISSUE THREAD

**Issue by colemcampbell** [realm/realm-cocoa#3677]
Swift 3 is removing implicit bridging conversions, which will cause issues with Realm's use of AnyObject as a parameter type for some methods. Here are two common examples:

## init(value:)

Take the following model:

```
class DatabaseObject: Object {
    dynamic var id: Int = 1
    dynamic var name: String = ""

    override class func primaryKey() -> String? {
        return "id"
    }
}
```

Currently, the following works (if you have imported Foundation):

`let databaseObject = DatabaseObject(value: [25, "object name"])`

However, it only works because the Array being passed, which in pure swift would be of type: `[Any]`, is being implicitly bridged to type `[NSObject]`. In Swift 3, this will no longer be the case.

This is part of a larger issue that Realm still doesn't feel completely like a native Swift API. For example, `init(value: AnyObject)` is strange in Swift, as it's meant to take either a Dictionary or an Array, but in Swift both are value types, not reference types, and therefore don't conform to AnyObject. At a minimum, the initializer should be `init(value: CollectionType)` (although in Swift 3 `CollectionType` will be renamed to `Collection`). This still would not necessarily be taking full advantage of Swift's type safety, however, because an array or dictionary could be passed that contains values that are of a type unsupported by Realm for model properties. This would cause a runtime error, but with Swift this can be a compile-time error instead.

One potential solution is for Realm to extend each type supported as a model property (Bool, Double, Float, etc) and add conformance to a protocol named something like "RealmObjectProperty". Then the initializer would be available with overloading as follows:

```
init(value: [RealmObjectProperty])
init(value: [String: RealmObjectProperty])
```

Not only will this work with Swift 3, but it is type-safe and eliminates some preventable run-time errors.

## objectForPrimaryKey(_:, key:)

Assuming the same example model as before:

`let databaseObject = realm.objectForPrimaryKey(DatabaseObject.self, key: 1)`

This will also fail to work in Swift 3 due to the removal of implicit bridging. This method has the same type issues as the other example. A primary key can be an Int, but Int does not conform to AnyObject, which is the type this method expects. This could be fixed by adding another protocol named "RealmObjectPrimaryKey" which inherits from RealmObjectProperty, then adding conformance to this protocol on String and Int. Then the function would be defined as:

`objectForPrimaryKey<T: Object>(type: T.Type, key: RealmObjectPrimaryKey) -> T?`

While this is probably out of scope for this issue, I just want to mention that this method is named more like an Objective-C method and will be not be in line with the Swift 3 API Guidelines. It should probably be named something like this:

`object(_:, withPrimaryKey:)`

In use:

`realm.object(Message.self, withPrimaryKey: 1)`

## Version of Realm and Tooling

Realm version: 1.0.0

Xcode version: 7.3.1 (7D1014)

Swift version: 2.2

**Comment by jpsim** [comment]
Thanks for taking the time to write this @colemcampbell! We've started looking at Swift 3 [very recently](https://github.com/realm/realm-cocoa/tree/jp-swift-3) and will need to reconsider many parts of the API, not just the interactions with `AnyObject`. You bring up naming, which will be another big one.

Your suggestion to add a `RealmObjectProperty` protocol that all supported Realm types conform to is a good idea that we'll be sure to consider, but might not be enough to support the current use cases. The initializer is designed to accept KVC-compliant input values, for example, that wouldn't be covered by this approach.

We're approaching adding Swift 3 support in 4 phases:

1. Make Realm Swift & its tests compile under the latest Swift 3 snapshot. (done)
2. Make all unit tests pass. (in progress)
3. Update API naming to reflect Swift 3 API guidelines.
4. Identify high friction areas of the API and improve them.

I'd like to keep this issue open to provide a place where we can discuss potential Swift 3 API design directions, but please keep in mind that we're early in the process.

**Comment by hartbit** [comment]
@jpsim What's the status on steps 3 and 4 now that Swift 3 is approximately 6 weeks from release? What about Swift 2.3 support?

**Comment by jpsim** [comment]
@hartbit we're maintaining support for the latest Xcode betas for both Swift 2.3 & 3.0 on our `master` branch. Step 3 was mostly done a few weeks ago by @austinzheng and @JadenGeller, though we'll want to review those changes and handle step 4 a bit closer to the Xcode 8 GM release.

**Comment by jpsim** [comment]
Finalized Swift 3 APIs were shipped in 1.1.0. See https://realm.io/news/realm-objc-swift-1.1.0/
